primary/src/block_synchronizer/peers.rs

```diff
@@ -0,0 +1,419 @@
+            .collect();
+
+        Peer {
+            name,
+            values_able_to_serve: certs,
+            assigned_values: HashMap::new(),
+        }
+    }
+
+    pub fn assign_values(&mut self, certificate: &Value) {
+        self.assigned_values
+            .insert(certificate.digest(), certificate.clone());
+    }
+
+    pub fn assigned_values(&self) -> Vec<Value> {
+        self.assigned_values
+            .clone()
+            .into_iter()
+            .map(|v| v.1)
+            .collect()
+    }
+}
+
+/// A helper structure to allow us store the peer result values
+/// and redistribute the common ones between them evenly.
+pub struct Peers<PublicKey: VerifyingKey, Value: Hash + Clone> {
+    /// A map with all the peers assigned on this pool.
+    peers: HashMap<PublicKey, Peer<PublicKey, Value>>,
+
+    /// When true, it means that the values have been assigned to peers and no
+    /// more mutating operations can be applied
+    rebalanced: bool,
+
+    /// An rng used to shuffle the list of peers
+    rng: SmallRng,
+}
+
+impl<PublicKey: VerifyingKey, Value: Hash + Clone> Peers<PublicKey, Value> {
+    pub fn new(rng: SmallRng) -> Self {
+        Self {
+            peers: HashMap::new(),
+            rebalanced: false,
+            rng,
+        }
+    }
+
+    pub fn peers(&self) -> HashMap<PublicKey, Peer<PublicKey, Value>> {
+        self.peers.to_owned()
+    }
+
+    /// Iterates over all the peer responses and retrieves the unique
+    /// values (identified by their digest).
+    pub fn unique_values(&self) -> Vec<Value> {
+        let result: HashMap<<Value as Hash>::TypedDigest, Value> = self
+            .peers
+            .values()
+            .flat_map(|v| {
+                if self.rebalanced {
+                    v.assigned_values.clone()
+                } else {
+                    v.values_able_to_serve.clone()
+                }
+            })
+            .collect();
+
+        result.into_iter().map(|(_, c)| c).collect()
+    }
+
+    pub fn contains_peer(&mut self, name: &PublicKey) -> bool {
+        self.peers.contains_key(name)
+    }
+
+    pub fn add_peer(&mut self, name: PublicKey, available_values: Vec<Value>) {
+        self.ensure_not_rebalanced();
+        self.peers
+            .insert(name.clone(), Peer::new(name, available_values));
+    }
+
+    /// Re-distributes the values to the peers in a load balanced manner.
+    /// We expect to have duplicates across the peers. The goal is in the end
+    /// for each peer to have a unique list of values and those lists to
+    /// not differ significantly, so we balance the load.
+    /// Once the peers are rebalanced, then no other operation that is allowed
+    /// mutates the struct is allowed.
+    pub fn rebalance_values(&mut self) {
+        self.ensure_not_rebalanced();
+
+        let values = self.unique_values();
+
+        for v in values {
+            self.reassign_value(v);
+        }
+
+        self.rebalanced = true;
```

Comment:
Yes, I haven't designed the `peers` structure to be thread-safe since I am expecting this to be modified by single thread a time anyways (e.x see `wait_for_payload_availability_responses` method, I don't have cases where 2 threads would be able to mutate the structure in the same time, or when the rebalancing happens is always from a single point). Although adding a thread-safety flavour to it might be good.